### PR TITLE
feat(storage): log transient errors retried on the sync paths

### PR DESCRIPTION
**Pull request by v-pratap**
## Problem

When GCS returns a transient error (429, 503, 500, `DEADLINE_EXCEEDED`) the
client backs off and retries with **no output at all**. From the application's
point of view a read just takes a long time.

There is nothing to grep for, so an operator cannot tell a slow network from a
server that is shedding load. The only way to see the retries today is to enable
the `rawclient` tracing component and pay for a full dump of every request and
response — which you can only do *after* you already suspect retries, i.e. not
during the first occurrence of an incident.

This came out of a real incident where reads on a large fleet appeared to stall
and the client produced nothing to correlate against.

## What this does

One `WARNING` record per **retried** attempt, nothing on success:

```text
[gcs-retry] <where>: transient error on attempt <n>, resource=<bucket/object>, status=<status>
```

This PR covers the **synchronous** paths, which serve both the JSON/REST and the
gRPC transports — GCS gRPC sync also goes through `RestRetryLoop`, because
`GenericStub` takes a `RestContext&` regardless of transport.

| Call site | `where` | Fires when |
|---|---|---|
| `ConnectionImpl::ReadObject` (retry lambda) | `ReadObject/open` | an attempt failed to open the stream |
| `RetryObjectReadSource::Read` | `ReadObject/resume` | the stream broke mid-download and is resumed from the last received byte |
| `ConnectionImpl::UploadChunk` | `UploadChunk` | an attempt to upload a chunk of a resumable upload failed |

New helper `google/cloud/storage/internal/retry_logging.{h,cc}`
(`LogTransientRetry`, `LoggingRetryPolicy`, `RetryLogResource`,
`RetryLogUploadResource`) so the call sites stay one line each.

Also makes `LoggingStub::ReadObject` report the status on failure. It cannot use
`LogWrapper` because `unique_ptr<ObjectReadSource>` has no `operator<<`, so it
previously logged nothing about *why* the call failed.

## Design notes

<details>
<summary><b>Why WARNING, and why not behind a tracing component</b></summary>

`LogSink::minimum_severity_` defaults to `TRACE`, but the default
`StdClogBackend` drops everything below `FATAL`
([log.cc#L228-L231](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/log.cc#L228-L231)).
So this is **invisible in a default OSS build** and becomes visible to anyone
who installs their own `LogBackend`. A diagnostic that has to be turned on first
is not useful during the first occurrence of an incident, which is exactly when
it is needed.

No changes to `google/cloud/internal/**` or `log.cc`; everything is inside
`google/cloud/storage/**`.
</details>

<details>
<summary><b>The status is flattened to one line</b></summary>

GCS error bodies end with a newline. Streaming the `Status` straight into
`GCP_LOG` carried that newline through, pushing the backend's `(file:line)`
suffix onto a **second line** — so `grep '\[gcs-retry\]'` returned a truncated
record, defeating the point of the greppable prefix.

It is also a small **log-injection** hazard, since the message is
service-supplied. `OneLine()` flattens `\n`, `\r`, `\t`.
Pinned by `RetryLogging.StatusIsFlattenedToASingleLine`.
</details>

<details>
<summary><b>The upload session URL is redacted</b></summary>

`UploadChunkRequest` only exposes `upload_session_url()`, and the `upload_id` in
it is a **bearer capability** — anyone holding it can append to or finalize that
upload with no other credential, for the lifetime of the session (~7 days).

The request's own `operator<<` does print it, but that is behind the opt-in
`rawclient` component, whereas this record is always on and lands in log
aggregators that are usually readable by more people than the bucket itself.

`RetryLogUploadResource()` keeps the bucket and an 8-character prefix of the id
— enough to tell concurrent uploads apart, not enough to hijack one.

Both transports are handled, because `UploadChunkRequest` carries whichever form
the transport produced: REST passes a URL with an `upload_id=` query parameter,
gRPC may pass `projects/<project>/buckets/<bucket>/<token>`, where the token may
itself contain `/`. Anything matching neither shape is **redacted whole**, on
the grounds that gRPC also issues wholly opaque tokens (see the testbench
section) and treating an unfamiliar value as safe would leak the entire
capability the first time the service changed format.
</details>

<details>
<summary><b>Every record comes from the retry policy in force</b></summary>

`ReadObject/open` has no natural place to count attempts: the work happens inside
a lambda that `RestRetryLoop()` calls. Rather than classifying the status
separately, it wraps the caller's policy in `LoggingRetryPolicy`, a
`google::cloud::RetryPolicy` decorator whose `OnFailure()` asks the wrapped
policy first and logs only when the answer is "retry".

A record is therefore emitted only when another attempt really does follow. A
permanent error, the final attempt of an exhausted policy, and a non-idempotent
request are all silent — each is already reported to the caller through the
returned `Status`. There is also no second copy of "which errors are retryable"
to drift away from the policy, so a caller who overrides `RetryPolicyOption`
automatically gets matching logs.

`ReadObject/resume` and `UploadChunk` already sat after their policy's
`OnFailure()`, so all three sites are now exact.

The backoff delay is deliberately not logged: reading it requires
`backoff_policy.OnCompletion()`, which *advances* the policy.
</details>

## Testbench verification

End-to-end against
[storage-testbench](https://github.com/googleapis/storage-testbench) `v0.61.0`
using its fault-injection API (`POST /retry_test`, id passed as
`x-retry-test-id` on REST and as call metadata on gRPC), with
`GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes`. Re-run in full after switching to the
`LoggingRetryPolicy` decorator. Each scenario asserts an **exact** record count.

| # | Scenario | Fault injected | Expected | Result |
|---|---|---|---|---|
| 1 | REST read, policy exhausted | `objects.get: 503 ×3`, 2 retries | 2 | PASS |
| 2 | REST read, zero-retry policy | `objects.get: 503`, 0 retries | 0 | PASS |
| 3 | REST read, retries succeed | `objects.get: 503 ×2`, 2 retries | 2 | PASS |
| 4 | REST read, permanent error | `objects.get: 400` | 0 | PASS |
| 5 | **Control** — healthy read | none | 0 | PASS |
| 6 | REST resumable upload | `objects.insert: return-503-after-256K` | 1 | PASS |
| 7 | gRPC resumable upload | `objects.insert: return-503-after-256K` | 1 | PASS |

Scenarios 1 and 2 are the ones the decorator changed. Previously the give-up
attempt was reported as though it were retried, and a policy granting no
retries still produced a record.

```text
===== 1. REST read, policy exhausted (3x503, 2 retries) -- give-up attempt must be silent
  [WARNING] [gcs-retry] ReadObject/open: transient error on attempt 1, resource=retry-logging-demo-bucket/small.txt, status=UNAVAILABLE: {"error":{"code":"503","message":{"error":{"message":"Retry Test: Caused a 503"}}}} (google/cloud/storage/internal/retry_logging.cc:144)
  [WARNING] [gcs-retry] ReadObject/open: transient error on attempt 2, resource=retry-logging-demo-bucket/small.txt, status=UNAVAILABLE: {"error":{"code":"503","message":{"error":{"message":"Retry Test: Caused a 503"}}}} (google/cloud/storage/internal/retry_logging.cc:144)
  --> 2 record(s), expected 2  [PASS]

===== 2. REST read, zero-retry policy (1x503, 0 retries) -- must be silent
  --> 0 record(s), expected 0  [PASS]

===== 4. REST read, permanent error -- must be silent
  --> 0 record(s), expected 0  [PASS]

===== 5. CONTROL: healthy REST read -- must be silent
  --> 0 record(s), expected 0  [PASS]

===== 6. REST resumable upload -- REST upload_id redacted
  [WARNING] [gcs-retry] UploadChunk: transient error on attempt 1, resource=http://localhost:9000/upload/storage/v1/b/retry-logging-demo-bucket/o?uploadType=resumable&upload_id=a40d2c89...[redacted], status=UNAVAILABLE: {"error":{"code":503,"message":"Fault injected during a resumable upload"}} (google/cloud/storage/internal/retry_logging.cc:144)
  --> 1 record(s), expected 1  [PASS]

===== 7. gRPC resumable upload -- opaque session token redacted
  [WARNING] [gcs-retry] UploadChunk: transient error on attempt 1, resource=7d7ab233...[redacted], status=UNAVAILABLE: {"error": {"errors": [{"domain": "global", "message": "Retry Test: Caused a StatusCode.UNAVAILABLE..."}]}} (google/cloud/storage/internal/retry_logging.cc:144)
  --> 1 record(s), expected 1  [PASS]
```

The trailing `(retry_logging.cc:144)` is added by the library's own log backend
(`operator<<(std::ostream&, LogRecord const&)` in `log.cc`), not by this change.

> [!IMPORTANT]
> Scenario 7 is why this section exists. The gRPC upload session id is a bare
> opaque 64-character hex string, not the `projects/_/buckets/<bucket>/<token>`
> resource name. The earlier code returned unrecognised shapes unchanged and so
> logged the whole token — a bearer credential — in clear. Redaction is now
> fail-closed: anything not structurally recognised is redacted whole, keeping
> only a short prefix for correlation. No unit test caught this.

## Tests

`bazel test //google/cloud/storage:all` → **175 / 175 PASSED** on this branch
alone.

New coverage in `retry_logging_test.cc`, `connection_impl_object_test.cc`,
`retry_object_read_source_test.cc`, `logging_stub_test.cc` — including negative
tests asserting **nothing** is logged on success, on a permanent error, on the
final attempt of an exhausted policy, or on a non-idempotent request.

`UploadChunkLogsTransientFailures` now uses a realistic resumable session URL
and asserts the bucket survives while the `upload_id` does not.

## Follow-ups

- The async and bidi paths are handled in a stacked PR on top of this one.
- Write coverage here is deliberately partial: only `UploadChunk`. A
  `RetryLoggingStub` `GenericStub` decorator installed in
  `generic_stub_factory.cc` would cover every unary RPC from a single place;
  worth doing separately rather than growing this PR.
- Log volume was raised in review and settled: one line per retried attempt is
  kept. `ExponentialBackoffPolicy` already spaces attempts exponentially in
  wall-clock time, and the decorator bounds the count at exactly the number of
  attempts the policy granted.

Internal ref: b/547926601

### CODE REVIEW

File: google/cloud/storage/internal/retry_logging.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_RETRY_LOGGING_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_RETRY_LOGGING_H
+
+#include "google/cloud/status.h"
+#include "google/cloud/version.h"
+#include "absl/strings/string_view.h"
```
[the remark this answers is not in the document]
Comment (v-pratap) [reply google/cloud/storage/internal/retry_logging.h]: Done. Switched to `#include <string_view>`.

Confirmed this is correct for the repo rather than just following the suggestion blindly: `.bazelrc` sets `-std=c++17` for linux and macos, and the style guide records `std::string_view` as a deliberate variance from upstream Abseil usage. `absl::StrCat` still accepts the views because Abseil aliases `absl::string_view` to `std::string_view` under C++17, so no call-site churn was needed.

File: google/cloud/storage/internal/retry_logging.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ * Formats a resumable upload session URL for the `resource` argument above.
+ *
+ * A resumable upload session URL carries an `upload_id`, and that id is a
+ * bearer capability: anyone who has it can append to, or finalize, the upload
+ * without any other credential. These diagnostics are always on and usually
+ * end up in a log aggregator that more people can read than the bucket itself,
+ * so only a short prefix of the id is kept. That is enough to correlate the
+ * log lines belonging to one upload, and not enough to hijack it.
+ */
+std::string RetryLogUploadResource(absl::string_view upload_session_url);
```
[the remark this answers is not in the document]
Comment (v-pratap) [reply google/cloud/storage/internal/retry_logging.h]: Done — all four signatures (`LogTransientRetry`, `RetryLogResource` x2 params, `RetryLogUploadResource`) now take `std::string_view`.

File: google/cloud/storage/internal/retry_logging.cc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    GCP_LOG(WARNING) << kRetryLogPrefix << " " << where
+                     << ": transient error on attempt " << attempt
+                     << ", status=" << OneLine(status);
+    return;
+  }
+  GCP_LOG(WARNING) << kRetryLogPrefix << " " << where
+                   << ": transient error on attempt " << attempt
+                   << ", resource=" << resource
+                   << ", status=" << OneLine(status);
+}
```
[the remark this answers is not in the document]
Comment (v-pratap) [reply google/cloud/storage/internal/retry_logging.cc]: Done, both parts.

- Definitions now take `std::string_view`.
- `pos`, `start`, `end` and `kept` are now explicitly `std::size_t` instead of `auto`.

I kept `auto constexpr kUploadId = std::string_view{"upload_id="};` as-is, since the type is written explicitly on the RHS — the style guide allows `auto` in that case, and your own suggested snippet kept it too.

File: google/cloud/storage/internal/connection_impl.cc [hunk 4]
```diff
@@ -409,13 +410,31 @@ StatusOr<std::unique_ptr<ObjectReadSource>> StorageConnectionImpl::ReadObject(
...
             rest_internal::RestContext& context, Options const& options,
-            ReadObjectRangeRequest const& request) {
+            ReadObjectRangeRequest const& request)
+            -> StatusOr<std::unique_ptr<ObjectReadSource>> {
           context.AddHeader(kIdempotencyTokenHeader, token);
-          return self->stub_->ReadObject(context, options, request);
+          StatusOr<std::unique_ptr<ObjectReadSource>> result =
+              self->stub_->ReadObject(context, options, request);
+          ++*attempt;
+          // Unlike the resume loop, this runs before the retry policy is
```
Comment (kalragauri) [inline google/cloud/storage/internal/connection_impl.cc]: Regarding this comment here that "this runs before the retry policy is consulted, so we cannot tell whether another attempt follows": since RestRetryLoop accepts google::cloud::RetryPolicy& by reference and calls retry_policy.OnFailure(last_status) after checking idempotency and right before sleeper(), we should be able to tell whether another attempt follows without touching google/cloud/internal/. 

Wrapping retry_policy in a small local google::cloud::RetryPolicy decorator whose OnFailure() logs only when inner_.OnFailure(status) returns true will make ReadObject/open consistent with ReadObject/resume and UploadChunk.

File: google/cloud/storage/internal/connection_impl.cc [hunk 5]
```diff
@@ -773,6 +792,15 @@ StatusOr<QueryResumableUploadResponse> StorageConnectionImpl::UploadChunk(
         return RetryError(std::move(last_status), *retry_policy, __func__);
       }
 
+      // `UploadChunkRequest` does not carry the bucket or object name, only the
+      // resumable session URL. That URL is the only stable identifier for the
+      // upload, but its `upload_id` is a bearer capability, so it is redacted
+      // before it reaches the log.
+      std::string const resource = storage_internal::RetryLogUploadResource(
```
Comment (kalragauri) [inline google/cloud/storage/internal/connection_impl.cc:790]: This path is shared by both the REST and gRPC flows, where a gRPC upload_id has the form `projects/_/buckets/<bucket>/<upload_session_token`, not a `projects/_/buckets/<bucket>/<object>` resource name (see this unit test: https://github.com/googleapis/google-cloud-cpp/blob/4365822261b0724c7b7f5a863d8f7a552ea81fe5/google/cloud/storage/internal/grpc/configure_client_context_test.cc#L204).

Please update RetryLogUploadResource() (and its unit tests) to also redact the session token suffix after `projects/<project>/buckets/<bucket>/`.

File: google/cloud/storage/internal/retry_logging.cc [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+bool IsTransientError(Status const& status) {
+  if (status.ok()) return false;
+  return !storage::internal::StatusTraits::IsPermanentFailure(status);
+}
+
+std::string RetryLogResource(std::string_view bucket_name,
+                             std::string_view object_name) {
+  if (bucket_name.empty() && object_name.empty()) return {};
+  return absl::StrCat(bucket_name, "/", object_name);
```
Comment (kalragauri) [inline google/cloud/storage/internal/retry_logging.cc:72]: This method should return just bucket_name or object_name when one is empty.

File: google/cloud/storage/internal/retry_logging_test.cc
Comment (kalragauri) [inline google/cloud/storage/internal/retry_logging_test.cc:1]: Could we add test cases for the negative / edge paths in `ReadObject` retry logging?
1. Verify that a permanent error (e.g. `kNotFound`) produces no `[gcs-retry]` log lines.
2. Verify that when retries exhaust (e.g. with `LimitedErrorCountRetryPolicy(2)` and 3 failures), only attempt 1 and attempt 2 are logged as retries, but not the 3rd (final) failure.
3. Verify that a transient error on a non-idempotent read request does not log a retry warning.

**Review by kalragauri — COMMENTED** [review]
Regarding your open question on log volume, logging every retried attempt (once ReadObject/open is gated on retry_policy.OnFailure() returning true) is the right choice IMHO rather than sampling powers of two. IIUC, ExponentialBackoffPolicy (the default retry policy) already spaces consecutive attempts exponentially in wall-clock time.

File: google/cloud/storage/internal/connection_impl.cc (hunk 4 above)
Comment (v-pratap) [reply google/cloud/storage/internal/connection_impl.cc]: Good catch — you are right, and I have made the change.

I verified the mechanism before implementing it. `RestRetryLoopImpl::operator()` in `google/cloud/internal/rest_retry_loop.h` returns early for a non-idempotent request and then calls `retry_policy.OnFailure(last_status)` immediately before `sleeper(...)`. So a decorator does see exactly the retry decision, and nothing under `google/cloud/internal/` needs to change.

Added `storage_internal::LoggingRetryPolicy` in `internal/retry_logging.h`. It wraps a `google::cloud::RetryPolicy&`, forwards the three virtuals, and `OnFailure()` asks the wrapped policy first — it logs only when the wrapped policy returned `true`. `ReadObject` now builds one on the stack inside the `factory` lambda and hands it to `RestRetryLoop`.

This fixed a behaviour bug, not just the inconsistency. The previous version logged the last attempt of an exhausted policy, i.e. it printed "transient error, retrying" on the very attempt where the library gives up. Now three cases are silent and correctly so — permanent error, final attempt of an exhausted policy, and non-idempotent request — because each is already reported to the caller through the returned `Status`.

Since the decision now comes from the policy in force, I also deleted `IsTransientError()`. It was a second copy of "which errors are retryable" that could drift away from the policy, and a caller who overrides `RetryPolicyOption` now automatically gets logging that matches their policy.

New tests in `retry_logging_test.cc`: `LoggingRetryPolicy.LogsEachRetriedAttempt`, `.PermanentErrorIsSilent`, `.ExhaustingThePolicyIsSilentOnTheFinalAttempt`, `.ForwardsToTheWrappedPolicy`, `.FollowsTheWrappedPolicyForAborted`.

File: google/cloud/storage/internal/connection_impl.cc (hunk 5 above)
Comment (v-pratap) [reply google/cloud/storage/internal/connection_impl.cc:790]: Thank you for this one — it was a real leak, not a style nit, and I would not have found it on my own.

I followed the test you linked. It shows `projects/_/buckets/test-bucket/test-upload-id`, and the case just below it shows the token itself can contain `/` (`.../test/upload/id`). My code only looked for a REST `upload_id=` query parameter, so on the gRPC path the entire session token went into the WARNING verbatim. Worse, I had a test named `UploadResourcePassesThroughWhenThereIsNoUploadId` that asserted exactly that behaviour, and a doc comment claiming gRPC was "a no-op here". Both were wrong and both are gone.

`RetryLogUploadResource()` now handles both shapes:

- REST: redacts the value of the `upload_id=` parameter, stopping at `&` so any following parameters survive.
- gRPC: for `projects/<project>/buckets/<bucket>/<token>`, redacts everything after the bucket segment — the whole remainder, not just the next path segment, precisely because the token may contain `/`.

Either way the first 8 characters are kept: enough to correlate the lines belonging to one upload, not enough to resume or finalize it. A string matching neither shape is returned unchanged, and a bare `projects/_/buckets/<bucket>` with no trailing segment passes through since there is no token to hide.

New tests: `UploadResourceRedactsTheGrpcSessionToken`, `UploadResourceRedactsAGrpcTokenContainingSlashes`, `UploadResourceKeepsABareBucketResource`, `UploadResourcePassesThroughUnknownShapes`.

File: google/cloud/storage/internal/retry_logging.cc (hunk 6 above)
Comment (v-pratap) [reply google/cloud/storage/internal/retry_logging.cc:72]: Fixed. `RetryLogResource()` now returns whichever name is present when the other is empty, and an empty string when both are. The old code produced `"bucket/"` or `"/object"`, which is noise in a log line and would break anyone splitting the field on `/`.

Pinned by `RetryLogging.ResourceOmitsTheSeparatorWhenOneNameIsEmpty`, alongside the existing `ResourceEmptyWhenUnset`. `LogTransientRetry()` already drops the whole `resource=` field when the string is empty, so the both-empty case stays clean too.

File: google/cloud/storage/internal/retry_logging_test.cc
Comment (v-pratap) [reply google/cloud/storage/internal/retry_logging_test.cc:1]: All three added. They live in `internal/connection_impl_object_test.cc` rather than `retry_logging_test.cc`, because each one has to drive the real `StorageConnectionImpl` retry loop to be meaningful:

1. `ReadObjectDoesNotLogPermanentFailures` — a `kNotFound` from the stub, asserts `IsEmpty()` on the captured `[gcs-retry]` records.
2. `ReadObjectLogsOnlyTheRetriedAttempts` — `RetryTestOptions()` allows 3 attempts; with 3 `UNAVAILABLE` failures the log holds exactly 2 records, for attempts 1 and 2, and nothing for the final failure.
3. `ReadObjectDoesNotLogNonIdempotentRequests` — a transient error on a non-idempotent read stays silent.

One note on (3): it needs a small local `NonIdempotentReads : public AlwaysRetryIdempotencyPolicy` override, because every shipped policy treats reads as idempotent — `StrictIdempotencyPolicy::IsIdempotent(ReadObjectRangeRequest const&)` returns `true` unconditionally. With that policy in place `RestRetryLoop` returns before it ever reaches `OnFailure()`, so the decorator is never consulted and nothing is logged.

Worth saying that (2) and (3) both failed against the previous version of this PR and pass only because of the decorator you suggested in the other thread. They were good tests to ask for.

**Comment by v-pratap** [comment]
Thanks — agreed, and I have dropped the throttling idea. Your reasoning holds up when I check it against the code: `ExponentialBackoffPolicy::OnCompletion()` scales the delay by `scaling_` each time, and the storage defaults are 1s initial / 5min maximum / 2x. So the records are already self-limiting in wall-clock terms; the interval between them grows until it hits the 5 minute ceiling. Sampling on top of that would mostly hide the early attempts, which are the ones an operator wants when deciding whether a spike is service-side.

The decorator you asked for in the `ReadObject/open` thread also bounds the total directly: the count is now exactly the number of attempts the policy granted, never more. Under `LimitedErrorCountRetryPolicy(n)` that is at most `n` lines per operation, and the final give-up attempt is silent. The old version could emit `n+1`.

So the behaviour is unchanged from what is in the PR body — one `WARNING` per retried attempt — and I have removed the open question about powers-of-two sampling.

**Comment by v-pratap** [comment]
Update: I re-ran this against the storage-testbench, and it found something that every unit test missed. Flagging it because it is security relevant and it extends an already-answered thread.

### The gRPC upload id was being logged in full

In the earlier discussion about redacting the upload id, the shape under consideration was the resource name `projects/<project>/buckets/<bucket>/<token>`. My `RetryLogUploadResource()` handled that shape and returned anything unrecognised **unchanged**.

Against a real gRPC resumable upload, the session id turned out to be a bare opaque 64-character hex string with no structure at all, so it fell into the unchanged branch and the log published it whole:

```
[gcs-retry] UploadChunk: transient error on attempt 1,
  resource=d8bec938b31cdd41225a65f98da14d12ada93fa6ef3f5ccfbc8840d29b828391,
  status=UNAVAILABLE: ...
```

That is the entire bearer capability for the upload session. The default was fail-open for a credential, which is the worst way to be wrong.

### Fix

`RetryLogUploadResource()` is now fail-closed: anything not structurally recognised is redacted whole, keeping only a short prefix so records from one upload can still be correlated. Same output against the testbench now:

```
[gcs-retry] UploadChunk: transient error on attempt 1,
  resource=7d7ab233...[redacted], status=UNAVAILABLE: ...
```

REST is unchanged and still keeps the useful part of the URL:

```
resource=http://localhost:9000/upload/storage/v1/b/retry-logging-demo-bucket/o
  ?uploadType=resumable&upload_id=a40d2c89...[redacted]
```

The header comment now explains why the default is fail-closed, so it does not get "helpfully" relaxed later. The unit test that asserted the pass-through behaviour has been deleted and replaced with tests for the opaque shapes, and `UploadChunkLogsTransientFailures` now uses a realistic resumable session URL and asserts that the bucket survives while the `upload_id` does not.

Full testbench matrix is in the PR description.

**Review by kalragauri — APPROVED** [review]
